[ExternalNode]Add ExternalNodeController for processing ExternalNode changes

mengdie-song opened the pull request:
1. Antrea Controller watches ExternalNode CRUD and converts it to
the corresponding ExternalEntity.

2. The conversion includes the following changes:
a.These ExternalEntity name is generated by this format:
  If NetworkInterface name is empty, uses ExternalNode name
  If NetworkInterface name is not empty, uses [ExternalNode name]-[Interface name].
b.The ExternalNode's labels are added on the ExternalEntity(ies).
c.The ExternalNode name is used to set "ExternalNode" field in the ExternalEntity.
d.ExternalNode NetworkInterface is used to set "Endpoint" field in the ExternalEntity.

3. Handle ExternalNode reconciliation and cleanup stale ExternalEntiies.

Signed-off-by: Mengdie Song <songm@vmware.com>

mengdie-song commented:
Hi @jianjuns @tnqn @wenyingd,

I have added an admission webhook in this change. 
The background is that currently there is no way to distinguish the ExternalEntity generated by ExternalNodeController and Users.
As a result, if users make some change(delete ExternalNode/update ExternalNode interfaces) when Antrea controller is down, we cannot clean up the stale ExternalEntities when the controller comes back.

Possible solutions we can think about right now are:
1. Add a webhook to prevent users from making ExternalNode changes(Update/Delete) when Antrea controller is down. (I have implemented in this PR)
2. Maybe add another field in ExternalEntity to distinguish the ExternalEntity generated by us automatically. (not sure if we like it)

I would like to know you ideas about it or maybe other better resolutions, thanks.

tnqn commented:
@mengdie-song this situation is usually resolved by adding a owner reference to externalentity, see an example here: https://github.com/kubernetes/kubernetes/blob/537941765fe1304dd096c1a2d4d4e70f10768218/pkg/controller/endpointslice/utils.go#L139-L146
Then you know which externalentity is created by antrea-controller and which externalNode it is created for.

mengdie-song commented:
@tnqn Thanks so much for the solution! I will update the change accordingly.

mengdie-song commented:
/test-all

jianjuns commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
```
ErrorS

Please change other code too

mengdie-song replied:
I think this one cannot use ErrorS as there is no error generated here. But I will update all the others.

wenyingd commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+			})
+		}
+	} else {
+		endpoints = append(endpoints, v1alpha2.Endpoint{
```
Why do we allow an ExternalEntity with empty IPs?

mengdie-song replied:
My original thought is to support interface name alone, interface IP alone and both name and IPs. Do we prefer to let user always specify IP?

wenyingd replied:
For ExternalNode case, yes, I think IP should be required.

mengdie-song replied:
OK, got it. Will update this part.

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+		return
+	}
+	if err := c.reconcileExternalNode(); err != nil {
+		klog.Errorf("Failed to reconcile ExternalNodes %v", err)
```
Thanks, got it. Will update.

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+	}
+	enNameSet := sets.NewString()
+	for _, en := range externalNodes {
+		enNameSet.Insert(en.Name)
```
Got it, will update it.

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+func (c *ExternalNodeController) addExternalEntity(ee *v1alpha2.ExternalEntity) error {
+	_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Create(context.TODO(), ee, metav1.CreateOptions{})
+	if errors.IsAlreadyExists(err) {
+		klog.Warningf("Updating ExternalEntity namespace %s, name %s as it already exists", ee.Namespace, ee.Name)
```
Got it. Will update.

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+	}
+	// Delete the previous ExternalEntity and create a new one if the name of the generated ExternalEntity is changed.
+	// Otherwise, update the ExternalEntity.
+	if preEn.Spec.Interfaces[0].Name != curEn.Spec.Interfaces[0].Name {
```
From my perspective,  we have already claimed that only interfaces[0] will be processed in the API definition. 
https://github.com/antrea-io/antrea/blob/feature/externalnode/pkg/apis/crd/v1alpha1/types.go#L638
So I prefer to make it simple here, what do your think？

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+
+func (c *ExternalNodeController) deleteExternalNode(namespace string, name string) error {
+	obj, exists, _ := c.syncedExternalNode.GetByKey(k8s.NamespacedName(namespace, name))
+	if !exists {
```
Thanks for the comment. I think this may happen when Add/Update and delete requests for the same ExternalNode are processed by two workers at the same time. I am thinking to change the worker number to one to avoid the potential concurrent issue, what do you think?

mengdie-song commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,384 @@
+	for _, ee := range externalEntities {
+		if len(ee.OwnerReferences) > 0 {
+			if !enNameSet.Has(ee.OwnerReferences[0].Name) {
+				klog.InfoS("Deleting stale ExternalEntity", "ExternalEntity", klog.KObj(ee))
```
Thanks, I have added Kind check for this in the latest commit.

mengdie-song commented:
/test-all
/test-windows-all
/test-ipv6-all
/test-ipv6-only-all

mengdie-song commented:
Hi @wenyingd @jianjuns @tnqn,

Could you help give another review on this change? Thanks.
I made another PR for AntreaAgentInfo https://github.com/antrea-io/antrea/pull/3757 which has dependency on the same externalNodeInformer defined in this change. I have also addressed the comments for ExternalNodeController part raised in https://github.com/antrea-io/antrea/pull/3757 in this PR.

wenyingd reviewed (approved):
LGTM

jianjuns commented on pkg/util/externalnode/externalnode.go:
```diff
@@ -0,0 +1,31 @@
+
+import "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+
+func GenExternalEntityName(externalNode *v1alpha1.ExternalNode) string {
```
Will it be used by other pkgs besides ExternalNodeController? If not I prefer to keep into the controller/externalnode or even controller.go for now.

jianjuns replied:
Ok, I saw you said it will be used by agent too. Then I am fine with the current way.

jianjuns commented on cmd/antrea-controller/controller.go:
```diff
@@ -155,6 +157,8 @@ func run(o *Options) error {
 		networkPolicyStore,
 		groupStore)
 
+	externalNodeController := externalnode.NewExternalNodeController(crdClient, externalNodeInformer, eeInformer)
```
Could we put the code under the external node feature gate?

mengdie-song replied:
OK, I will update this part and add a feature gate on controller side.

jianjuns commented on pkg/controller/externalnode/controller.go:
```diff
@@ -0,0 +1,375 @@
+		}
+		eeName := externalnode.GenExternalEntityName(en)
+		enUIDEENameMap[en.UID] = eeName
+	}
```
Should we add existing nodes to syncedExternalNode?

mengdie-song replied:
They will be added to syncedExternalNode inside function addExternalNode in line 151.

jianjuns replied:
Ok. I missed it.

jianjuns commented:
/test-all
